kotlinx-coroutines-core/common/src/sync/Semaphore.kt

```diff
@@ -121,76 +121,80 @@ private class SemaphoreImpl(private val permits: Int, acquiredPermits: Int) : Se
     }
 
     override suspend fun acquire() {
-        val p = _availablePermits.getAndDecrement()
-        if (p > 0) return // permit acquired
-        addToQueueAndSuspend()
+        while (true) {
+            val p = _availablePermits.getAndDecrement()
+            if (p > 0) return // permit acquired
+            if (addToQueueAndSuspend()) return
```

Comment:
This is not a tail-call call anymore. It should be refactored using fast-path/slow-path patter (where fast-path does not suspend and calls slow-path suspending method as a tail-call). See how it is done in channels, for example.